m3/src/main/java/com/uber/m3/tally/m3/M3Reporter.java

```diff
@@ -0,0 +1,781 @@
+            // thread(s) or else they will block forever
+            executor.shutdownNow();
+
+            // Wait a maximum of `MAX_PROCESSOR_WAIT_ON_CLOSE_MILLIS` for all processors
+            // to return from flushing
+            try {
+                phaser.awaitAdvanceInterruptibly(
+                    phaser.arrive(),
+                    MAX_PROCESSOR_WAIT_ON_CLOSE_MILLIS,
+                    TimeUnit.MILLISECONDS
+                );
+            } catch (TimeoutException e) {
+                LOG.warn(
+                    "M3Reporter closing before Processors complete after waiting timeout of %dms!",
+                    MAX_PROCESSOR_WAIT_ON_CLOSE_MILLIS
+                );
+            } catch (InterruptedException e) {
+                LOG.warn("M3Reporter closing before Processors complete due to being interrupted!");
+            }
+
+            transport.close();
+
+            isOpen = false;
+        }
+    }
+
+    private static Set<MetricTag> toMetricTagSet(Map<String, String> tags) {
+        Set<MetricTag> metricTagSet = new HashSet<>();
+
+        if (tags == null) {
+            return metricTagSet;
+        }
+
+        for (Map.Entry<String, String> tag : tags.entrySet()) {
+            metricTagSet.add(createMetricTag(tag.getKey(), tag.getValue()));
+        }
+
+        return metricTagSet;
+    }
+
+    private static MetricTag createMetricTag(String tagName, String tagValue) {
+        MetricTag metricTag = new MetricTag(tagName);
+
+        if (tagValue != null && !tagValue.isEmpty()) {
+            metricTag.setTagValue(tagValue);
+        }
+
+        return metricTag;
+    }
+
+    private int calculateSize(Metric metric) {
+        try {
+            synchronized (calcLock) {
+                metric.write(calcProtocol);
+
+                return calc.getSizeAndReset();
+            }
+        } catch (TException e) {
+            LOG.warn("Unable to calculate metric batch size. Defaulting to: " + DEFAULT_METRIC_SIZE);
+            return DEFAULT_METRIC_SIZE;
+        }
+    }
+
+    private String valueBucketString(double bucketBound) {
+        if (bucketBound == Double.MAX_VALUE) {
+            return "infinity";
+        }
+
+        if (bucketBound == -Double.MAX_VALUE) {
+            return "-infinity";
+        }
+
+        return String.format(bucketValFmt, bucketBound);
+    }
+
+    private String durationBucketString(Duration bucketBound) {
+        if (Duration.ZERO.equals(bucketBound)) {
+            return "0";
+        }
+
+        if (Duration.MAX_VALUE.equals(bucketBound)) {
+            return "infinity";
+        }
+
+        if (Duration.MIN_VALUE.equals(bucketBound)) {
+            return "-infinity";
+        }
+
+        return bucketBound.toString();
+    }
+
+    @Override
+    public void reportCounter(String name, Map<String, String> tags, long value) {
+        synchronized (isOpenLock) {
```

Comment:
Does this (`synchronized` on `isOpenLock`) affect throughput much?

Reply:
As discussed offline, removed the need for an `isOpen` flag. Instead, use a sentinel metric in the queue to indicate that all metrics coming after it was put in after reporter was closed.